File: api/src/main/java/net/blueshell/api/event/ContributionEventListener.java
package net.blueshell.api.event;

import net.blueshell.api.common.event.PostPersistEvent;
import net.blueshell.api.common.event.PostRemoveEvent;
import net.blueshell.api.common.event.PostUpdateEvent;
import net.blueshell.api.job.contact.AddContactToListJob;
import net.blueshell.api.job.contact.RemoveContactFromListJob;
import net.blueshell.api.model.contribution.Contribution;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.transaction.event.TransactionPhase;
import org.springframework.transaction.event.TransactionalEventListener;

@Component
public class ContributionEventListener {

    private final AddContactToListJob addContactToListJob;
    private final RemoveContactFromListJob removeContactFromListJob;

    public ContributionEventListener(AddContactToListJob addContactToListJob,
                                     RemoveContactFromListJob removeContactFromListJob) {
        this.addContactToListJob = addContactToListJob;
        this.removeContactFromListJob = removeContactFromListJob;
    }

    @TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
    @Transactional(propagation = Propagation.REQUIRES_NEW)
    public void onPersist(PostPersistEvent<Contribution> evt) {
        var c = evt.getSource();
        addContactToListJob.addToList(c.getUserId(), c.getContributionPeriodId());
    }

    @TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
    @Transactional(propagation = Propagation.REQUIRES_NEW)
    public void onUpdate(PostUpdateEvent<Contribution> evt) {
        var c = evt.getSource();
        addContactToListJob.addToList(c.getUserId(), c.getContributionPeriodId());
    }

    @TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
    @Transactional(propagation = Propagation.REQUIRES_NEW)
    public void onDelete(PostRemoveEvent<Contribution> evt) {
        var c = evt.getSource();
        removeContactFromListJob.removeFromList(c.getUserId(), c.getContributionPeriodId());
    }
}

File: api/src/main/kotlin/net/blueshell/api/contribution/application/listener/ContributionEventListener.kt
package net.blueshell.api.contribution.application.listener

import net.blueshell.api.contribution.application.event.ContributionChangeType
import net.blueshell.api.contribution.application.event.ContributionChangedEvent
import net.blueshell.api.platform.integration.event.job.AddContactToListEvent
import net.blueshell.api.platform.integration.event.job.RemoveContactFromListEvent
import org.springframework.context.ApplicationEventPublisher
import org.springframework.context.event.EventListener
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Propagation
import org.springframework.transaction.annotation.Transactional

@Component
class ContributionEventListener(
    private val eventPublisher: ApplicationEventPublisher
) {
    @EventListener
    @Transactional(propagation = Propagation.REQUIRES_NEW)
    fun onChange(evt: ContributionChangedEvent) {
        when (evt.changeType) {
            ContributionChangeType.CREATED,
            ContributionChangeType.UPDATED -> eventPublisher.publishEvent(
                AddContactToListEvent(evt.userId, evt.periodId)
            )
            ContributionChangeType.DELETED -> eventPublisher.publishEvent(
                RemoveContactFromListEvent(evt.userId, evt.periodId)
            )
        }
    }
}
